### ISSUE TITLE
Build OS image and store it on provisioner node

### ISSUE THREAD

**Issue by invidian** [tinkerbell/cluster-api-provider-tinkerbell#19]
To avoid having large `postKubeadmCommands` and `preKubeadmCommands` sections in Cluster template, we could follow Openstack provider https://image-builder.sigs.k8s.io/capi/providers/openstack.html#building-images-for-openstack and instruct user to build required OS image on the Tinkerbell provisioner, so it can be served using nginx running there.

This will simplify the installation process and make it faster.

**Comment by detiber** [comment]
+1, I'd love to have a way to build and store images for tinkerbell to use. I'm not sure if the right approach is to store them on the mirror host for direct PXE booting, or to take a container based approach like what is done with Equinix Metal, where we could potentially source the images from Artifact Hub similar to the published workflows that were recently announced.

@gianarb do you have thoughts here?

**Comment by invidian** [comment]
>+1, I'd love to have a way to build and store images for tinkerbell to use. I'm not sure if the right approach is to store them on the mirror host for direct PXE booting, or to take a container based approach like what is done with Equinix Metal, where we could potentially source the images from Artifact Hub similar to the published workflows that were recently announced.

@detiber I think mirroring is still valuable as it will decrease provisioning time and make it independent from external services (think offline scenarios). Having an Artifact Hub sounds like an interesting concept still.

**Comment by gianarb** [comment]
TLDR: Great idea, but let's start simpler; let's run Minio or push OS to an s3 bucket and see how it goes.

---

Proposal 0021 that speaks about Actions, mentions the idea to reconfigure the registry we ship as part of the Tinkerbell Stack to proxy the images used by the action we "certify." I think it is the right way to go, and we can probably do something similar for OS. ArtifactHub is just a metadata representation at the end of the story; the asset has to be pushed and managed elsewhere for actions we use quay.io for now.

We should wait for @rawkode to work a bit at the infrastructure repository allowing the Tinkerbell Community to have its own infrastructure. After that, we can set up something like Nexus or Minio or an Nginx server to serve an operating system (and we can hack something to proxy it in the tink stack). If we like that and think it is useful, we can build metadata and push them to ArtifactHub. But for this particular scenario, I would like to wait a bit more before interacting with the ArtifactHub community because I want to validate and see how the action's implementation proceed.

**Comment by tstromberg** [comment]
@thebsdbox - does your work with Crocodile obsolete this issue?

**Comment by detiber** [comment]
@tstromberg agreed, I think the future approach will be to move towards the OCI streaming.

/close

**Comment by detiber** [comment]
Public images are now available from the github container registry and can be overridden through configuration to point to other registries.

There is no longer a need to try and figure out where to store images.
